src/CoreWCF.Primitives/src/CoreWCF/Dispatcher/SecurityServiceDispatcher.cs

```diff
@@ -0,0 +1,593 @@
+
+        public ICollection<Type> SupportedChannelTypes => InnerServiceDispatcher.SupportedChannelTypes;
+
+        public object ThisLock { get; } = new object();
+
+        public SecurityProtocolFactory SecurityProtocolFactory
+        {
+            get
+            {
+
+                //  ThrowIfDisposed();
+                return this.securityProtocolFactory;
+            }
+            set
+            {
+                if (value == null)
+                {
+                    throw DiagnosticUtility.ExceptionUtility.ThrowHelperArgumentNull("value");
+                }
+                //   ThrowIfDisposedOrImmutable();
+                this.securityProtocolFactory = value;
+            }
+        }
+
+        public bool SessionMode
+        {
+            get
+            {
+                return this.sessionMode;
+            }
+            set
+            {
+                this.sessionMode = value;
+            }
+        }
+
+        public SecuritySessionServerSettings SessionServerSettings
+        {
+            get
+            {
+                if (this.sessionServerSettings == null)
+                {
+                    lock (ThisLock)
+                    {
+                        if (this.sessionServerSettings == null)
+                        {
+                            SecuritySessionServerSettings tmp = new SecuritySessionServerSettings();
+                            System.Threading.Thread.MemoryBarrier();
+                            tmp.SecurityServiceDispatcher = this;
+                            this.sessionServerSettings = tmp;
+                        }
+                    }
+                }
+                return this.sessionServerSettings;
+            }
+        }
+
+        bool SupportsDuplex
+        {
+            get
+            {
+                ThrowIfProtocolFactoryNotSet();
+                return this.securityProtocolFactory.SupportsDuplex;
+            }
+        }
+
+        bool SupportsRequestReply
+        {
+            get
+            {
+                ThrowIfProtocolFactoryNotSet();
+                return this.securityProtocolFactory.SupportsRequestReply;
+            }
+        }
+
+        public bool SendUnsecuredFaults
+        {
+            get
+            {
+                return this.sendUnsecuredFaults;
+            }
+            set
+            {
+                this.sendUnsecuredFaults = value;
+            }
+        }
+
+        public IChannel OuterChannel
+        {
+            get { return this.outerChannel; }
+            set { this.outerChannel = value; }
+        }
+
+        IList<Type> IServiceDispatcher.SupportedChannelTypes => throw new NotImplementedException();
+
+        public ServiceHostBase Host =>  InnerServiceDispatcher.Host;
+
+        //private 
+
+        // This method should only be called at Open time, since it looks up the identity based on the 
+        // thread token
+        //void ComputeEndpointIdentity()
+        //{
```

Comment:
This will become important when we add eventually add support for Kerberos authentication with Message Credentials.

Reply:
Have you added this in your Windows Auth implementation? There are 2 ways to handle the server identity with Nego, one is the client explicitly provides it, the other is the server provides it. I don't know whether the latter uses this or has the SSPI layer do it as part of the Nego handshake. I will at least need this when I add WSDL support.